prov/util/src/util_mem_monitor.c

```diff
@@ -146,49 +146,80 @@ void ofi_monitors_cleanup(void)
 	memhooks_monitor->cleanup(memhooks_monitor);
 }
 
-int ofi_monitor_add_cache(struct ofi_mem_monitor *monitor,
-			  struct ofi_mr_cache *cache)
+/* Monitors array must be of size OFI_HMEM_MAX. */
+int ofi_monitors_add_cache(struct ofi_mem_monitor **monitors,
+			   struct ofi_mr_cache *cache)
 {
 	int ret = 0;
+	enum fi_hmem_iface iface;
+	struct ofi_mem_monitor *monitor;
 
-	if (!monitor)
+	if (!monitors) {
+		for (iface = FI_HMEM_SYSTEM; iface < OFI_HMEM_MAX; iface++)
+			cache->monitors[iface] = NULL;
 		return -FI_ENOSYS;
+	}
 
 	pthread_mutex_lock(&mm_lock);
-	if (dlist_empty(&monitor->list)) {
-		if (monitor == uffd_monitor)
-			ret = ofi_uffd_start();
-		else if (monitor == memhooks_monitor)
-			ret = ofi_memhooks_start();
-		else
-			ret = -FI_ENOSYS;
-
-		if (ret)
-			goto out;
+
+	for (iface = FI_HMEM_SYSTEM; iface < OFI_HMEM_MAX; iface++) {
+		monitor = monitors[iface];
+		if (!monitor) {
+			FI_DBG(&core_prov, FI_LOG_MR,
+			       "MR cache disabled for %s memory\n",
+			       fi_tostr(&iface, FI_TYPE_HMEM_IFACE));
+			cache->monitors[iface] = NULL;
+			continue;
+		}
+
+		if (dlist_empty(&monitor->list)) {
+			if (monitor == uffd_monitor)
+				ret = ofi_uffd_start();
+			else if (monitor == memhooks_monitor)
+				ret = ofi_memhooks_start();
```

Comment:
Do we need a start() callback, so we can avoid checks like this?

Reply:
Will add it in.